stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,829 @@
+#ifndef STAN_MATH_REV_FUNCTOR_INTEGRATE_ODE_CVODES_ADJOINT_HPP
+#define STAN_MATH_REV_FUNCTOR_INTEGRATE_ODE_CVODES_ADJOINT_HPP
+
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/core/save_varis.hpp>
+#include <stan/math/rev/functor/cvodes_utils.hpp>
+#include <stan/math/rev/functor/ode_store_sensitivities.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/value_of.hpp>
+#include <stan/math/prim/functor/for_each.hpp>
+#include <cvodes/cvodes.h>
+#include <nvector/nvector_serial.h>
+#include <sunmatrix/sunmatrix_dense.h>
+#include <sunlinsol/sunlinsol_dense.h>
+#include <algorithm>
+#include <ostream>
+#include <utility>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Integrator interface for CVODES' ODE solvers (Adams & BDF
+ * methods).
+ *
+ * @tparam F Type of ODE right hand side
+ * @tparam T_y0 Type of scalars for initial state
+ * @tparam T_param Type of scalars for parameters
+ * @tparam T_t0 Type of scalar of initial time point
+ * @tparam T_ts Type of time-points where ODE solution is returned
+ */
+template <typename F, typename T_y0, typename T_t0, typename T_ts,
+          typename... T_Args>
+class cvodes_integrator_adjoint_vari : public vari {
+  using T_Return = return_type_t<T_y0, T_t0, T_ts, T_Args...>;
+  using T_y0_t0 = return_type_t<T_y0, T_t0>;
+
+  const std::decay_t<F> f_;
+  arena_t<Eigen::Matrix<T_y0_t0, Eigen::Dynamic, 1>> y0_;
+  arena_t<T_t0> t0_;
+  std::vector<arena_t<T_ts>, arena_allocator<arena_t<T_ts>>> ts_;
+
+  double relative_tolerance_forward_;
+  arena_t<Eigen::VectorXd> absolute_tolerance_forward_;
+  double relative_tolerance_backward_;
+  arena_t<Eigen::VectorXd> absolute_tolerance_backward_;
+  double relative_tolerance_quadrature_;
+  double absolute_tolerance_quadrature_;
+  long int max_num_steps_;                  // NOLINT(runtime/int)
+  long int num_steps_between_checkpoints_;  // NOLINT(runtime/int)
+  int interpolation_polynomial_;
+  int solver_forward_;
+  int solver_backward_;
+
+  std::ostream* msgs_;
+
+  std::tuple<arena_t<
+      plain_type_t<decltype(deep_copy_vars(std::declval<const T_Args&>()))>>...>
+      local_args_tuple_;
+  std::tuple<arena_t<
+      plain_type_t<decltype(value_of(std::declval<const T_Args&>()))>>...>
+      value_of_args_tuple_;
+
+  size_t N_;
+  bool forward_returned_;
+  bool backward_is_initialized_;
+
+  size_t num_t0_vars_;
+  size_t num_ts_vars_;
+  size_t num_y0_vars_;
+  size_t num_args_vars_;
+  size_t num_vars_;
+
+  arena_t<Eigen::VectorXd> state_forward_;
+  arena_t<Eigen::VectorXd> state_backward_;
+  arena_t<Eigen::VectorXd> quad_;
+
+  vari** non_chaining_varis_;
+
+  vari** t0_varis_;
+  vari** ts_varis_;
+  vari** y0_varis_;
```

Comment:
Since we store these in the vars for each of these (if they exist) in the arena now we should be able to get rid of these pointers

Reply:
but how would be access the adjoints? The upside of using the pointers is that these are a single memory block such that access should be fast.

Reply:
Let me take a crack at what I'm thinking of and if we don't like it we won't do it

Reply:
So I'm thinking about something like this

https://github.com/stan-dev/math/compare/feature/adjoint-odes...review/ode-adjoint4

There's some finesse required to get rid of the args varis but if you like the above I think I can sort that out

Reply:
Accessing the values of t0 etc should still have linear memory access, but now we don't have to duplicate the number of vari in the class.

The args part will be a bit weirder just because we need to replace 

```cpp
    if (is_any_var_args_) {
      for (size_t s = 0; s < num_args_vars_; ++s) {
        args_varis_[s]->adj_ += quad_.coeff(s);
      }
    }
```

with something that updates the vars in the tuple of args iteratively 

```cpp
    if (is_any_var_args_) {
      update_adjoints(local_args_tuple_, quad);
    }
```

Reply:
And if I can sort that out here we can probably also sort out how to do it in `reduce_sum()` etc. other places where we use this pattern

Reply:
I need to look in more detail into this. The code looks neat, but I am not sure if we actually have the time now to refactor this bit right now... let me look one more time.

Just one thing: We do not want to update local_args_tuple_. That one is a local deep copy of the args. We need to modify the args_tuple passed into the vari. Currently we don't save that one since we manipulate only the args_varis_.

So we would get rid of a fully copy of the varis when doing this?

Reply:
Kind of a dumb q but why do we make a local copy of the args? Took a shot at this today but I must be missing something as I can't get it to work

Reply:
Because we stick these args into the nested ad calls repeatedly. This avoids creating copies for each evaluation (the usual way), but it also changes their adjoints such that we have to have a dedicated single copy.

Reply:
icic, okay then idt we should do this for the args yet, but I think the rest makes sense. The alt here _I think_ is to 

1. Make `local_args_tuple` the actual args
2. store the adjoints in a tuple before doing those nested calls and set the adjoints to zero in `local_args_tuple_`
3. do  the nested ad calls
4. Reapply the adjoint values we saved before to `local_args_tuple_` adjoints

We would use 1/3rd as much memory (since each element of the vari** is pointer + two doubles), but it feels v complicated so I'm fine with not doing this for the args right now. 

Though take a look at the other ones because I think it makes sense for those.

Reply:
This save and restore approach is something I implemented once and threw it away. It's really messy to do this. It's also not too thread-safe :D . So you suggest to take over the other stuff in ode-adjoint4? I will go over it to digest it then... but really, we should stop then there and just complete the tests. The release deadline is coming close and this still needs a review...

Reply:
Yeah +1 on all fronts. That change feels really messy and there's no reason we can't do it in the future if we want to

Reply:
Just merged ode-adjoints4. Tests are passing locally for me so as long as that works we should be good. Though if you want to wait to review it I can revert it